tests/components/home_connect/test_coordinator.py

```diff
@@ -0,0 +1,319 @@
+from homeassistant.core import HomeAssistant
+
+from tests.common import MockConfigEntry
+
+
+async def test_coordinator_update(
+    config_entry: MockConfigEntry,
+    integration_setup: Callable[[MagicMock], Awaitable[bool]],
+    setup_credentials: None,
+    client: MagicMock,
+) -> None:
+    """Test that the coordinator can update."""
+    assert config_entry.state == ConfigEntryState.NOT_LOADED
+    await integration_setup(client)
+    assert config_entry.state == ConfigEntryState.LOADED
+
+    assert cast(HomeConnectConfigEntry, config_entry).runtime_data.data
+
+
+async def test_coordinator_update_failing_get_appliances(
+    config_entry: MockConfigEntry,
+    integration_setup: Callable[[MagicMock], Awaitable[bool]],
+    setup_credentials: None,
+    client_with_exception: MagicMock,
+) -> None:
+    """Test that the coordinator raises ConfigEntryNotReady when it fails to get appliances."""
+    client_with_exception.get_home_appliances.return_value = None
+    client_with_exception.get_home_appliances.side_effect = HomeConnectError()
+
+    assert config_entry.state == ConfigEntryState.NOT_LOADED
+    await integration_setup(client_with_exception)
+    assert config_entry.state == ConfigEntryState.SETUP_RETRY
+
+
+async def test_coordinator_update_failing_get_settings_status(
+    config_entry: MockConfigEntry,
+    integration_setup: Callable[[MagicMock], Awaitable[bool]],
+    setup_credentials: None,
+    client_with_exception: MagicMock,
+) -> None:
+    """Test that although is not possible to get settings and status, the config entry is loaded.
+
+    This is for cases where some appliances are reachable and some are not in the same configuration entry.
+    """
+    # Get home appliances does pass at client_with_exception.get_home_appliances mock, so no need to mock it again
+    assert config_entry.state == ConfigEntryState.NOT_LOADED
+    await integration_setup(client_with_exception)
+    assert config_entry.state == ConfigEntryState.LOADED
+
+    assert cast(HomeConnectConfigEntry, config_entry).runtime_data.data
+
+
+@pytest.mark.parametrize(
+    ("event_type", "event_key"),
+    [
+        (EventType.STATUS, EventKey.BSH_COMMON_STATUS_DOOR_STATE),
+        (EventType.NOTIFY, EventKey.BSH_COMMON_SETTING_POWER_STATE),
+        (EventType.EVENT, EventKey.DISHCARE_DISHWASHER_EVENT_SALT_NEARLY_EMPTY),
+    ],
+)
+async def test_event_listener(
+    event_type: EventType,
+    event_key: EventKey,
+    hass: HomeAssistant,
+    config_entry: MockConfigEntry,
+    integration_setup: Callable[[MagicMock], Awaitable[bool]],
+    setup_credentials: None,
+    client: MagicMock,
+    appliance_ha_id: str,
+) -> None:
+    """Test that the event listener works."""
+    assert config_entry.state == ConfigEntryState.NOT_LOADED
+    await integration_setup(client)
+    assert config_entry.state == ConfigEntryState.LOADED
+
+    listener = MagicMock()
+    coordinator = cast(HomeConnectConfigEntry, config_entry).runtime_data
```

Comment:
We should still remove the direct use of the coordinator here. I'd instead test a specific entity and check that the entity state is updated as long as the entity is added to Home Assistant and isn't updated after the entity is removed. An entity can be removed via the entity registry API.

Reply:
Got it. But if I remove the entity, then how do I check that the entity doesn't receive any updates (so that it keeps the same state)

Reply:
Instead of removing the entity I think we can change the entity_id of the entity in the entity registry. This will remove the entity and then add it again. If we wouldn't clean up stale listeners there would be two listeners attached to the entity after the change of entity_id. We can listen for `EVENT_STATE_REPORTED` and check that the state isn't updated again with the same state.

Check that the logic checks out by commenting out the clean up and the test should fail.